Lib/asyncio/proactor_events.py

```diff
@@ -426,6 +427,120 @@ def _pipe_closed(self, fut):
+        # constructor does it for us.
+        self._buffer = collections.deque()
+        self._loop.call_soon(self._loop_reading)
+
+    def _set_extra(self, sock):
+        super()._set_extra(sock)
+        self._extra['socket'] = sock
+
+        try:
+            self._extra['sockname'] = sock.getsockname()
+        except (socket.error, AttributeError):
+            if self._loop.get_debug():
+                logger.warning(
+                    "getsockname() failed on %r", sock, exc_info=True)
+
+        if 'peername' not in self._extra:
+            try:
+                self._extra['peername'] = sock.getpeername()
+            except (socket.error, AttributeError):
+                if self._loop.get_debug():
+                    logger.warning("getpeername() failed on %r",
+                                   sock, exc_info=True)
+
+
+    def abort(self):
+        self._force_close(None)
+
+    def sendto(self, data, addr=None):
+        if not isinstance(data, (bytes, bytearray, memoryview)):
+            raise TypeError('data argument must be byte-ish (%r)',
+                            type(data))
+
+        if not data:
+            return
+
+        if self._conn_lost and self._address:
+            # close() or force_close() has been called on the bound endpoint
+            return
+
+        self._buffer.appendleft((data, addr))
+
+        if self._write_fut is None:
+            # No current write operations are active, kick one off
+            self._loop_writing()
+        else:
+            # A write operation is already kicked off
+            pass
+
+    def _loop_writing(self, fut=None):
+        if self._conn_lost:
+            return
+
+        assert fut is self._write_fut
+        if fut:
+            # We are in a _loop_writing() done callback, get the result
+            fut.result()
+
+        if not self._buffer or (self._conn_lost and self._address):
+            # The connection has been closed
+            self._write_fut = None
+            return
+
+        data, addr = self._buffer.pop()
+
+        self._write_fut = None
+        try:
+            if self._address:
+                self._write_fut = self._loop._proactor.send(self._sock, data)
+            else:
+                self._write_fut = self._loop._proactor.sendto(self._sock, data, addr=addr)
+        except OSError as exc:
+            self._protocol.error_received(exc)
+            self._fatal_error(exc, 'Fatal error sending UDP datagram')
+        else:
+            self._write_fut.add_done_callback(self._loop_writing)
+
+    def _loop_reading(self, fut=None):
+        if self._conn_lost:
+            return
+
+        assert self._read_fut is fut
+
+        if fut:
+            res = fut.result()
+
+            if self._address:
+                data, addr = res, self._address
+            else:
+                data, addr = res
+
+            self._protocol.datagram_received(data, addr)
```

Comment:
Again, see the other _loop_reading(): it seems like data must be ignored if we are closing the socket ("if self._closing:").